### PR TITLE
Data pipeline

### PR DESCRIPTION
**Pull request by Ranapop**
This PR creates a first draft of the data pipeline:
* the sentences and amrs with alignements are read from the data files (see definitions.py for data paths)
* the amrs are parsed from string format using the penman package
* the amr in penman format (with the alignments) is used to extract the list of ordered concepts and the adjacency matrix
* a pytorch dataset is implemented, together with a collate function that can be used for batching
  * each batch is a dict of sentences, concepts and adjacency matrices, padded with zeros to have the same shape in the batch and stacked

For now an example of how the dataset can be used is given for the training data and can be run from the project root with:
`python -m data_pipeline.dataset`

This will be removed once more tests are added & the dataset will be used in training.

This PR does not contain:
* obtaining an AMR string from pytorch tensors (concepts & adj mat)

### CODE REVIEW

File: data_pipeline/concepts.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def get_ordered_concepts_ids(g: penman.Graph, unalignment_tolerance = 0):
+  """
+  Constructs a list of ordered concepts.
+
+  Args:
+    g: amr graph (in penman library representation).
+    unalignment_tolerance: the percentage of concepts with no alignemnt info
+      allowed. If the amr has a larger percantage than allowed trhough this
```
Comment (alex-frasie) [inline data_pipeline/concepts.py]: Typos: percentage (I think), through

File: data_pipeline/concepts.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  unaligned_concepts = [c for c in concept_ids if c not in aligned_concepts]
+  no_unaligned_concepts = len(unaligned_concepts)
+  if no_unaligned_concepts/no_concepts > unalignment_tolerance:
+    return None
+  # Only use the first alignment for sorting (for now).
+  result = sorted(aligned_concepts, key = lambda c: concepts_alignments[c][0])
+  # Place unaligned concepts randomly.
+  for c in unaligned_concepts:
+    random_index = 0
+    if len(result) != 0:
```
Comment (alex-frasie) [inline data_pipeline/concepts.py:82]: But what would happen if len(result) would be 0? 
We would then generate the value randomly with randrange(0). Would that fail? Wouldn't it give 0 automatically?

File: data_pipeline/adjacency_mat.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import penman
+
+INSTANCE_ARC_LABEL = ":instance"
+ROOT_LABEL = ':root'
+
+
+def get_concepts_to_triple_ids(g: penman.Graph):
+  """
+  Extract a dictionary concept -> triple id (where the 
+  concept is instanced).
```
Comment (alex-frasie) [inline data_pipeline/adjacency_mat.py]: Why the new line 😆 or why here, in the middle of the paranthesis?

File: data_pipeline/data_reading.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  # Read file data into a string.
+  file = open(path,'r')
+  text = file.read()
+  file.close()
+  # Separate the data for each amr (id, token, alignemnts, amr).
+  data_blocks = text.split('\n\n')
+  triples = []
+  for data_block in data_blocks:
+    # Use regex to extract the needed info from a text block.
+    pattern = r'# ::id(.*)\n# ::tok(.*)\n# ::alignments(.*)\n(?=\()(.*)'
```
Comment (alex-frasie) [inline data_pipeline/data_reading.py]: This looks scary 😃 Maybe we can talk a little bit about this one.
Also can we extract this as a costant?

File: data_pipeline/adjacency_mat.py (hunk 3 above)
Comment (Ranapop) [reply data_pipeline/adjacency_mat.py]: typo

File: data_pipeline/data_reading.py (hunk 4 above)
Comment (Ranapop) [reply data_pipeline/data_reading.py]: Extracted it as a constant and put a comment to be easier understood.

File: data_pipeline/concepts.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    A list of the triple ids (positions in the original triples list) for the
+    triples associated with concepts (that have a concept on the 3rd position). 
+  """
+  triples = g.triples
+  variables = [t[0] for t in triples]
+  variables = list(set(variables))
+  filtered_triples_ids = []
+  for triple_id in range(len(triples)):
+    t = triples[triple_id]
+    src = t[0]
```
Comment (teodora99) [inline data_pipeline/concepts.py]: This is not used. Is there a reason you declared it?

File: data_pipeline/concepts.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
+from random import randrange
+from typing import List, Dict
+
+import penman
+from penman.graph import Triple
```
Comment (teodora99) [inline data_pipeline/concepts.py]: I think this import is not used.

**Review by alex-frasie — APPROVED** [review]

File: data_pipeline/concepts.py (hunk 5 above)
Comment (Ranapop) [reply data_pipeline/concepts.py]: Removed now

File: data_pipeline/concepts.py (hunk 6 above)
Comment (Ranapop) [reply data_pipeline/concepts.py]: That's correct. Will remove

**Review by georgimarian — APPROVED** [review]

**Pull request merged**